### ISSUE TITLE
Use a sevice worker to improve low-bandwidth/offline experience

### ISSUE THREAD

**Issue by jankeromnes** [jankeromnes/janitor#4]
Implementing a service worker in the Janitor web app could significantly improve subsequent load times, and make it (somewhat) usable offline.

**Comment by jankeromnes** [comment]
A basic service worker was added in 6468bdac21f6aa07b99060ba5f1d3760a564586b, but it doesn't provide caching yet.

**Comment by Coder206** [comment]
@jankeromnes Would this be similar to what you are looking for: https://serviceworke.rs/strategy-cache-update-and-refresh.html or perhaps https://github.com/GoogleChrome/sw-precache ?

**Comment by jankeromnes** [comment]
Yes, we're looking to achieve similar results, but we don't have a build system to hook into (and we probably won't have, as HTTP 2 will remove the need to bundle resources).

Additionally, we might need custom caching logic, because resources like `/img/*.svg` should always be cached, but API calls like `/$spawn` or the upcoming `/api/hosts` should often not be cached, and we should handle offline requests to these differently. I'm not sure yet what the best way to deal with API calls with a Service Worker is.

**Comment by nt1m** [comment]
@jankeromnes I don't think this is essential for the beta release tbh.

**Comment by jankeromnes** [comment]
@nt1m I'm worried that the more we work on the website without writing a service worker, the harder it will become to eventually write one. But you're right, it's not a hard requirement and it shouldn't block the beta release. Let's remove the issue from the milestone.

**Comment by jankeromnes** [comment]
@Coder206 found this, which could help: https://www.pwabuilder.com/generator
